### DISCUSSION TITLE
How to keep the send function in one's context?

### DISCUSSION THREAD

**Discussion by github0013** [statelyai/xstate#2268] [Q&A] [answered]
Let's say I had parent - child machines.

1. the parent has one child service object in its context
2. the child service has the parent's `send` function in its context
3. so that I can call the parent's events from the child (it's like propagating an event from bottom to top)

## parent

```tsx
createMachine(
  {
    context: {
      child: childService,
    },
...
...
```

## child

```tsx
createMachine(
  {
    context: {
      parentSend: ...,
    },
...
...
```

Can this be done without actually creating the parent's service object then splitting the `send` function to insert into child's context? 
(I don't want the user to manually apply parent's send function AFTER creating the services)

The only way I can think of is something like this

1. have a init state
2. invoke a service
3. use the callback as parent's send
4. initialize child's context with it

```tsx
createMachine<Context, Event>(
  {
    initial: "init",
    context: {
      child: undefined,
    },
    states: {
      init: {
        invoke: {
          id: "init",
          src: "initialize",
        },
        on: {
          ASSIGN_CHILD: {
            actions: ["assignChild"],
          },
          DONE: {
            target: "idle",
          },
        },
      },
      idle: {
        on: {
          NOTIFY: {
            actions: ["notify"],
          },
        },
      },
    },
  },
  {
    services: {
      initialize: (context, event) => (callback, onReceive) => {
        const childService = interpret(
          child.withContext({ parentSend: callback })
        ).start()

        callback({ type: "ASSIGN_CHILD", child: childService })
        callback({ type: "DONE" })
      },
    },
    actions: {
      assignChild: assign({
        child: (context, event) => {
          if (event.type !== "ASSIGN_CHILD") return context.child

          return event.child
        },
      }),
      notify: () => {
        console.log("notify: from parent")
      },
    },
  }
)
```

Is there a better way? or it's just a bad practice?

**Comment by davidkpiano** [comment, marked as answer]
You can spawn a machine instead:

```js
init: {
  entry: assign({
    child: () => spawn(childMachine)
  })
}
```

And in the `childMachine` you would send events to the parent using the `actions.sendParent(event)` action creator.
